File: sheet/prog/hash.py
#!/usr/bin/env python3

from collections import defaultdict, OrderedDict

# python ではdictと呼ばれる
person = {
    "name": "Kyohei, Takahashi",
    "nickname": "kcrt",
    "addr": "kcrt@kcrt.net",
    "age": 34
}
my_dict = dict(one=1, two=2, three=3)   # これでも作れる

person["nickname"]  # => "kcrt"
# person["nokey"]   # [[ERROR]] KeyError
person.get("nokey", "default")   # なければ引数で指定したもの(省略時None)を返す (例外にならない)

person["birthday"] = "1984-09-24"   # 辞書に追加・変更
del person["age"]                   # 辞書から削除
my_item = person.pop("addr")        # 取得して削除
len(person)         # => 4 : 項目の数

"age" in person     # キーが含まれているかの確認

# 結合
dict1 = {"a": 1, "b": 2}
dict2 = {"a": 3, "d": 4}
dict1.update(dict2)     # dict1 += dict2的な
print(dict1)

# 走査
for (key, value) in person.items():
    print(f"{key} is {value}")
# もしくは
for key in person.keys():
    print(f"{key} is {person[key]}")
# もしくは
for value in person.values():
    print(f"{value}")

# 内包表記
# 配列と同じく内包表記が便利
# { キー: 値 for 仮引数 in 対象のリスト if 条件 } (if 条件は省略可能)
myperson = {name: value.upper() for (name, value) in person.items()}

# defaultdict
mydict = defaultdict(int)
print(mydict["a"])  # => 0 : エラーにならない
for c in ["aa", "bb", "cc"]:
    mydict[c] += 1  # 初めてのアクセス時に要素が出来る

# defaultdict(int)のintは型名でなく関数
# アイテムがなければint() すなわち 0をセットする
# defaultdict(lambda: random.randint(1, 10))なんて事も出来る

# OrderedDict
# defaultdictと違い、ordereddictでないのはこちらの方が新しいから(see: naming_rule)
mydict = OrderedDict()
mydict["a"] = 1
mydict["b"] = 1
mydict["c"] = 1
mydict.keys()    # => 必ず追加した順になる
# 順番が意味を持つjsonファイルなどを扱う際に便利(see: json)

File: sheet/prog/hash.jl
#!/usr/bin/env julia

# Dict
person = Dict(
	"name" => "Kyohei, Takahashi",
	"nickname" => "kcrt",
	"addr" => "kcrt@kcrt.net",
	"age" => 34
)
println(person)
keytype(person)		# 型を取得
valtype(person)	# 
# 型は推測される(上記だとDict{String, Any})
# {"a" => 1, "b" => 2} という書き方があったが[[obsolete]]

# tupleの配列でもよい。
# また、型を指定することも出来る。
my_dict = Dict{String, Int64}([("a", 1), ("b", 2)])
Dict(:a => 1, :b => 2)	# keyにシンボルを使用することも出来る。
# データの取得
println(person["nickname"])	# => "kcrt"
# person["nokey"]	# [[ERROR]] KeyError
get(person, "nokey", "default")	# 無ければデフォルトを返す
get!(person, "nokey", "default")	# 無ければデフォルトを格納

# 辞書に追加
person["birthday"] = "1984-09-24"
# 辞書から削除
delete!(person, "age")
# 取得して削除
my_item = pop!(person, "age", "default")	# 取得して削除(defaultは省略可)

# 項目数の取得
println(person)
println("length: ", length(person))

# キーの有無の確認
haskey(person, "name")
println("name" in keys(person))

# 結合
dict1 = Dict("a" => 1, "b" => 2)
dict2 = Dict("a" => 3, "d" => 4)
d = merge(dict1, dict2)		# 結合Dictを返す
println(d)
# または、
merge!(dict1, dict2)		# dict1に結合
println(dict1)

merge!(+, dict1, dict2)		# Keyが被っている場合は和にする
println(dict1)

# 走査
for (key, value) in pairs(person)	# =, in, ∈ のどれでもOK
    println("$key is $value")
end
for key in keys(person)	# =, in, ∈ のどれでもOK
    println(key)
end
for value in values(person)	# =, in, ∈ のどれでもOK
    println(value)
end
# 内包表記
myperson = Dict(key => uppercase(value) for (key, value) in person)
println(myperson)
# defaultdict, ordereddict

# fruits["Ehime"]		# "=> Mikan": 値の取得
# fruits["Kagawa"] = "Udon"	# 値の設定
# get(fruits, "Hiroshima", "Lemon")	# 取得(デフォルト付き)
# 
# 
# keys(fruits)	# キーの集合
# values(fruits)	# 値の集合


# Pair
mypair = "a" => 1	# Pair{String,Int64}
mypair[1] # => "a"
mypair[2] # => 1

